HW-1 네이버 영화 검색

nukeolaf opened the pull request:
네이버 영화 검색 API 를 사용하여 영화 검색 어플리케이션을 만들었습니다

kutta97 committed:
메인 레이아웃에 SearchView 추가

kutta97 committed:
gradle 에 리사이클러뷰 dependency 추가

kutta97 committed:
gradle 리포맷

kutta97 committed:
메인 레이아웃에 리사이클러뷰 추가

kutta97 committed:
메인 레이아웃 constraint -> linear 으로 수정

kutta97 committed:
영화 목록 리사이클러뷰에 사용할 아이템 리소스 추가

kutta97 committed:
gradle 에 cardview dependency 추가

kutta97 committed:
cardview 로 영화 리사이클러뷰 아이템뷰 만들기

kutta97 committed:
영화 아이템 뷰에서 사용하지 않는 constraint guideline 제거

kutta97 committed:
영화 리사이클러뷰 레이아웃에 listitem attribute 에 아이템 추가

kutta97 committed:
gradle 에 retrofit dependency 추가

kutta97 committed:
RetrofitInterface 추가

kutta97 committed:
RetrofitInterface 에 영화 검색하는 GET 함수 추가

kutta97 committed:
네이버 지도 api 사용을 위한 client id 와 client secret 추가함

kutta97 committed:
searchView 레이아웃에 id 추가

kutta97 committed:
영화 검색 결과 출력 recyclerView 레이아웃에 id 추가

kutta97 committed:
searchView 에 검색 이벤트 리스너 추가

kutta97 committed:
recyclerView height 을 0dp 로 바꾸고, weight 속성 제거함

kutta97 committed:
retrofit client object 클래스 추가

kutta97 committed:
retrofit 인터페이스를 MainActivity 전역변수로 선언

kutta97 committed:
retrofit 인터페이스를 onCreate 에서 초기화

kutta97 committed:
inputString 의 공백확인 메서드 추가

kutta97 committed:
검색시 검색어가 공백이 아닌경우에 영화 검색 api 호출하는 코드 추가

kutta97 committed:
RetrofitInterface searchMovie() 에서 다중쿼리 map 을 <String, Any> 로 수정

kutta97 committed:
영화 검색시 검색어가 공백일때 다이얼로그 띄워주는 함수 추가

kutta97 committed:
영화 검색시 검색어가 공백일때 showEmptyDialog() 호출

kutta97 committed:
영화 목록 출력 리사이클러뷰 어댑터 클래스 작성

kutta97 committed:
영화 api 응답 data class "Movie", "Item" 추가

kutta97 committed:
Retrofit Interface searchMovie() 의 리턴 Call<Movie> 추가

kutta97 committed:
HashMap value 값의 Any 때문에 생긴 오류 수정
RetrofitInterface GET 함수에서 쿼리 맵 사용하지 않음
Parameter type must not include a type variable or wildcard: java.util.Map<java.lang.String, ?>

kutta97 committed:
인터넷 사용 퍼미션 추가

kutta97 committed:
UserRating 자료형을 Float 으로 수정

kutta97 committed:
레트로핏 통신 응답 body 출력 로그 추가

kutta97 committed:
Movie data class 에서 안 쓰이는 변수 제거

kutta97 committed:
data class 이름 MovieItem, MovieResponse 로 수정

kutta97 committed:
data class 이름 MovieItem, MovieResponse 로 수정

kutta97 committed:
MovieAdapter 파라미터로 영화 목록 리스트 받을 수 있게 함

kutta97 committed:
MovieAdapter onCreateViewHolder 작성

kutta97 committed:
MovieAdapter getItemCount 작성

kutta97 committed:
MovieAdapter onBindViewHolder 작성

kutta97 committed:
MovieAdapter 에 ViewHolder 클래스 작성, onBindViewHolder() 작성

kutta97 committed:
gradle 에 Glide dependency 추가

kutta97 committed:
MovieAdapter 에서 view binding 할때 Glide 로 이미지 보여주는 코드 추가

kutta97 committed:
영화 검색하는 코드 searchMovie() 함수로 모듈화

kutta97 committed:
MainActivity onCreate 에서 영화 목록 리사이클러뷰에 어댑터 추가해줌

kutta97 committed:
Adapter 철자 오류 수정

kutta97 committed:
검색했을때 movieAdapter 로 검색결과 영화 목록 ArrayList 보여주는 코드 추가

kutta97 committed:
검색 결과 영화 보여주는 코드 showMovies() 함수로 모듈화

kutta97 committed:
Log 태그 수정

kutta97 committed:
리사이클러뷰 높이 wrap_content 로 수정

0dp 로 하면 아예 안보임

kutta97 committed:
리사이클러뷰 inflating 할때 attachToParent false 로 수정

kutta97 committed:
검색어 공백 입력 검사하는 함수와 다이얼로그 띄우는 함수 제거

kutta97 committed:
리사이클러뷰 아이템 레이아웃 height 을 wrap_content 로 수정

kutta97 committed:
리사이클러뷰 영화 아이템 리스너 추가

kutta97 committed:
영화 상세내용 보여줄 액티비티 추가

kutta97 committed:
MovieAdapter 에 MovieItemListener 추가

kutta97 committed:
영화 별점 표시하는 rating bar 크기 수정

kutta97 committed:
영화 아이템뷰 글씨 길어지면 뒤에 ... 으로 나오게 함

kutta97 committed:
영화 별점 표시하는 코드 추가

kutta97 committed:
movieAdapter 에 movieItemListener 인터페이스 객체 작성하여 추가

kutta97 committed:
movieItemListener 파라미터로 MovieItem 넘기도록 수정

kutta97 committed:
activity_movie_detail 레이아웃 작성

kutta97 committed:
별점 텍스트에 id 추가

kutta97 committed:
MovieDetailActivity 에서 intent 로 넘어온 MovieItem 보여주는 코드 추가

kutta97 committed:
MovieDetailActivity 영화 이미지 크기 수정

kutta97 committed:
영화 제목 뒤에 괄호로 제작 연도 붙여주기

kutta97 committed:
Html.fromHtml 을 사용해서 <b> 태그 없앰

kutta97 committed:
search_view 속성 추가해줌

kutta97 committed:
감독이랑 출연진 String 에 "|" 이거 없애줌

kutta97 committed:
html.ToString() 함수로 모듈화

eunjin-hong commented on 2004/nukeolaf/app/src/main/java/com/olaf/nukeolaf/MovieDetailActivity.kt:
```diff
@@ -18,8 +18,8 @@ class MovieDetailActivity : AppCompatActivity() {
-        val director = "감독 : ${movieItem.director}"
-        val actor = "출연진 : ${movieItem.actor}"
+        val director = "감독 : ${movieItem.director.replace("|", " ")}"
+        val actor = "출연진 : ${movieItem.actor.replace("|", " ")}"
```
🙋‍♀️ 외국 감독, 출연진들의 이름에는 띄어쓰기가 있으니, `|` 가 마음에 안 드신다면 다른 구분자를 넣어주시는건 어떨까요? (`,`나 `/` 등 )

nukeolaf replied:
수정했습니다! 03dd1de

eunjin-hong commented on 2004/nukeolaf/app/src/main/java/com/olaf/nukeolaf/MainActivity.kt:
```diff
@@ -52,8 +52,13 @@ class MainActivity : AppCompatActivity() {
-                Log.d("onQueryTextListener", response.body().toString())
+                movieAdapter.run {
+                    setMovies(response.body()!!.items)
+                    notifyDataSetChanged()
```
🙋‍♀️`!!` 대신 `response.body()` 가 null일 경우에 대응하면 더 좋을 것 같습니다.

nukeolaf replied:
수정했습니다!
7facaa9

eunjin-hong commented on 2004/nukeolaf/app/src/main/java/com/olaf/nukeolaf/MainActivity.kt:
```diff
@@ -37,6 +37,7 @@ class MainActivity : AppCompatActivity() {
                     ).enqueue(object : Callback<Movie> {
                         override fun onResponse(call: Call<Movie>, response: Response<Movie>) {
                             Log.d("onQueryTextListener", response.toString())
+                            Log.d("onQueryTextListener", response.body().toString())
```
🙋‍♀️[OkHttpLoggingInterceptor](https://github.com/square/okhttp/tree/master/okhttp-logging-interceptor) 를 이용하시면 매번 로그를 찍지 않아도 로그를 볼 수  있습니다!

nukeolaf replied:
꿀팁 알려주셔서 감사합니다~ㅎㅎ
수정했습니다! 
0550522

eunjin-hong commented on 2004/nukeolaf/app/src/main/java/com/olaf/nukeolaf/MainActivity.kt:
```diff
@@ -1,12 +1,74 @@
+    private lateinit var retrofitClient: RetrofitInterface
+    private val itemListener: MovieItemListener = object : MovieItemListener {
+        override fun onMovieItemClick(item: MovieItem) {
+            var intent = Intent(this@MainActivity, MovieDetailActivity::class.java)
```
🙋‍♀️저희 과제가  영화 상세 페이지를 인터넷 브라우저에서 띄우기였던거같아요!  확인 부탁드립니다!

ddaeyeonkim replied:
intent 변수 선언시 `val` 로 해주세요

nukeolaf replied:
@eunjin-hong 수정했습니다! abfea91
@improve777 수정했습니다! 63acd12

ddaeyeonkim commented on 2004/nukeolaf/app/src/main/java/com/olaf/nukeolaf/MainActivity.kt:
```diff
@@ -1,12 +1,74 @@
+            startActivity(intent)
+        }
+    }
+    private val movieAdapter = MovieAdapter(ArrayList(0), itemListener)
```
Adapter 생성자에 list 를 넣지 않도록 하는게 좋습니다

nukeolaf replied:
comment 감사합니다!
Adapter 생성자에 list 를 넣지 않도록 하는게 좋다는 것은
list 는 Adapter 의 전역변수로 선언하고, 나중에 set 함수로 초기화해주는 게 좋다는 뜻으로 이해해도 괜찮을까요?
일단은 제가 이해한대로 수정했습니다! bb4ae17

///아래에 코멘트가 접혀있던걸 이제 봤습니다! 죄송합니다~ㅜㅠ

nukeolaf replied:
앗..! 제가 생각한대로 수정했더니 에러가 떠서 수정했습니다!
5d78aff

ddaeyeonkim commented on 2004/nukeolaf/app/src/main/java/com/olaf/nukeolaf/MainActivity.kt:
```diff
@@ -1,12 +1,74 @@
+        movie_rv.adapter = movieAdapter
+
+        search_view.setIconifiedByDefault(false)
+        search_view.queryHint = "영화 검색"
```
xml 에도 해당 속성이 있긴 한데 xml 에서 적용이 안되서 코드에도 넣으신 것 같습니다.
xml 의 namespace 를 `android` 에서 `app` 으로 변경해주시면 위 코드는 생략이 가능할 것 같습니다.

nukeolaf replied:
수정했습니다! bef13e9

ddaeyeonkim commented on 2004/nukeolaf/app/src/main/res/layout/activity_main.xml:
```diff
@@ -1,18 +1,24 @@
-        app:layout_constraintRight_toRightOf="parent"
-        app:layout_constraintTop_toTopOf="parent" />
+        android:iconifiedByDefault="false"
+        android:queryHint="검색어를 입력하시오" />
```
이렇게 하면 속성이 적용되네요

```suggestion
        app:iconifiedByDefault="false"
        app:queryHint="검색어를 입력하시오" />
```

ddaeyeonkim commented on 2004/nukeolaf/app/src/main/java/com/olaf/nukeolaf/MovieAdapter.kt:
```diff
@@ -0,0 +1,70 @@
+
+    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
+
+        private var view = itemView
```
itemView 가 전역변수로 있기 때문에 따로 안만드셔도 될거같아요

nukeolaf replied:
수정했습니다! 8dd2b05

ddaeyeonkim commented on 2004/nukeolaf/app/src/main/java/com/olaf/nukeolaf/MovieAdapter.kt:
```diff
@@ -0,0 +1,70 @@
+            val director = "감독 : ${item.director.replace("|", " ")}"
+            val actor = "출연진 : ${item.actor.replace("|", " ")}"
+
+            view.apply {
```
```suggestion
            itemView.apply {
```

ddaeyeonkim commented on 2004/nukeolaf/app/src/main/java/com/olaf/nukeolaf/MovieAdapter.kt:
```diff
@@ -0,0 +1,70 @@
+                    itemListener.onMovieItemClick(item)
+                }
+            }
+            Glide.with(view)
+                .load(item.image)
+                .into(view.movie_image)
```
```suggestion
            Glide.with(itemView)
                .load(item.image)
                .into(itemView.movie_image)
```

ddaeyeonkim commented on 2004/nukeolaf/app/src/main/java/com/olaf/nukeolaf/MovieAdapter.kt:
```diff
@@ -0,0 +1,70 @@
+        val inflatedView = LayoutInflater
+            .from(parent.context)
+            .inflate(R.layout.item_movies_rv, parent, false)
+        return (ViewHolder(inflatedView))
```
바깥쪽 괄호는 지워주세요

nukeolaf replied:
수정했습니다!
ab0cd48

ddaeyeonkim commented on 2004/nukeolaf/app/src/main/java/com/olaf/nukeolaf/MovieAdapter.kt:
```diff
@@ -0,0 +1,70 @@
+) : RecyclerView.Adapter<MovieAdapter.ViewHolder>() {
+
+    fun setMovies(list: ArrayList<MovieItem>) {
+        this.movies = list
```
어댑터 내부에서 따로 list 를 선언하고 `add()`, `addAll()` 해서 관리하는게 좋습니다.

nukeolaf replied:
와...! 이런 방법이 있었군요!! 
알려주셔서 감사합니다!! 정말 많이 배워갑니다!!
말씀하신대로 수정했습니다! 
cd57a9b

ddaeyeonkim commented on 2004/nukeolaf/app/src/main/java/com/olaf/nukeolaf/MovieAdapter.kt:
```diff
@@ -0,0 +1,70 @@
+import kotlinx.android.synthetic.main.item_movies_rv.view.*
+
+class MovieAdapter(
+    private var movies: ArrayList<MovieItem>,
```
생성자에서 제거해주세요

nukeolaf replied:
수정했습니다! bb4ae17

ddaeyeonkim commented on 2004/nukeolaf/app/src/main/java/com/olaf/nukeolaf/MovieAdapter.kt:
```diff
@@ -0,0 +1,70 @@
+    private val itemListener: MovieItemListener
+) : RecyclerView.Adapter<MovieAdapter.ViewHolder>() {
+
+    fun setMovies(list: ArrayList<MovieItem>) {
```
리스트를 인자로 받을때는 `List<MovieItem>` 으로 받는게 좋습니다.

nukeolaf replied:
수정했습니다! f950740

ddaeyeonkim commented on 2004/nukeolaf/app/src/main/java/com/olaf/nukeolaf/MainActivity.kt:
```diff
@@ -1,12 +1,74 @@
+                call: Call<MovieResponse>,
+                response: Response<MovieResponse>
+            ) {
+                showMovies(response.body()!!.items).also {
```
response 에 대한 체크를 먼저 해주는게 좋습니다.

```kotlin
if ((response.body() != null) and response.isSuccessful) {
      // todo              
}
```

nukeolaf replied:
수정했습니다! ee849bc

ddaeyeonkim commented on 2004/nukeolaf/app/src/main/java/com/olaf/nukeolaf/MainActivity.kt:
```diff
@@ -1,12 +1,74 @@
+        })
+    }
+
+    private fun showMovies(movies: ArrayList<MovieItem>) {
```
여기도 `List<MovieItem>` 으로 선언해주시는게 좋습니다

nukeolaf replied:
수정했습니다! 94e555d

ddaeyeonkim commented on 2004/nukeolaf/app/src/main/java/com/olaf/nukeolaf/MovieAdapter.kt:
```diff
@@ -0,0 +1,70 @@
+
+        private var view = itemView
+
+        fun bind(item: MovieItem, position: Int) {
```
position 은 빼주셔도 될 것 같아요

nukeolaf replied:
수정했습니다! 35ffe72

ddaeyeonkim commented on 2004/nukeolaf/app/build.gradle:
```diff
@@ -25,11 +25,17 @@ android {
+    implementation 'com.squareup.retrofit2:retrofit:2.3.0'
+    implementation 'com.squareup.retrofit2:converter-gson:2.3.0'
+    implementation 'com.github.bumptech.glide:glide:4.10.0'
+    annotationProcessor 'com.github.bumptech.glide:compiler:4.10.0'
```
글라이드 공식문서에서는 코틀린 사용시 `kapt` 로 선언하라고 명시되어 있습니다.

nukeolaf replied:
수정했습니다! 25acab2

ddaeyeonkim commented on 2004/nukeolaf/app/src/main/res/layout/item_movies_rv.xml:
```diff
@@ -0,0 +1,117 @@
+                android:layout_marginEnd="10dp"
+                android:ellipsize="end"
+                android:maxLines="1"
+                android:text="감독 : 김아무개"
```
디자인할때만 참고할수 있도록 `tools:text` 로 작성해주시는게 좋습니다

nukeolaf replied:
수정했습니다! 1646cc7

ddaeyeonkim commented on 2004/nukeolaf/app/src/main/res/layout/activity_main.xml:
```diff
@@ -1,18 +1,24 @@
+    <androidx.recyclerview.widget.RecyclerView
+        android:id="@+id/movie_rv"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
```
recyclerview 는 `wrap_content` 로 안하시는게 좋습니다

nukeolaf replied:
수정했습니다! aed9ffe

ddaeyeonkim commented on 2004/nukeolaf/app/src/main/java/com/olaf/nukeolaf/MovieItem.kt:
```diff
@@ -0,0 +1,14 @@
+    val director: String,
+    val actor: String,
+    val userRating: Float
+) : Serializable
```
`Parcelable` 로 구현해주시는게 어떨까요?

nukeolaf replied:
수정했습니다! e7dae33

ddaeyeonkim commented on 2004/nukeolaf/app/src/main/java/com/olaf/nukeolaf/MovieResponse.kt:
```diff
@@ -0,0 +1,9 @@
+    val total: String,
+    val start: String,
+    val display: String,
+    val items: ArrayList<MovieItem>
```
여기도 `List<MovieItem>` 으로 선언해주시는게 좋습니다

nukeolaf replied:
수정했습니다! f950740
List<> 는 읽기만 가능한 자료구조라고 알고 있는데,
왜 List로 선언해주는게 좋은지 여쭤봐도 괜찮을까요???
검색해봐도 이유를 잘 못 찾겠습니다ㅠㅜ

// 이제 무슨 말씀이셨는지 알 것 같습니다!
코멘트 달아주셔서 감사합니다!
정말 많이 배웠습니다

ddaeyeonkim reviewed (commented):
아직은 코드량이 많지 않아서 괜찮지만 확장성을 고려해서 패키지 분리를 해주시는게 좋을 듯 합니다.

kutta97 committed:
intent 객체 var -> val 로 변경

kutta97 committed:
리사이클러뷰 아이템 클릭시, 영화 url 로 웹페이지 띄우도록 수정

kutta97 committed:
영화 상세 내용 보여주는 액티비티 제거
(과제 내용이랑 달라서 수정)

kutta97 committed:
레트로핏 응답이 성공했을때와 실패했을때 조건문 작성

kutta97 committed:
토스트 메시지 출력하는 함수 작성

kutta97 committed:
api 응답 response 의 상태에 따른 조건 처리

kutta97 committed:
레트로핏 BASE_URL 을 도메인으로 수정해줌

kutta97 committed:
gradle 에 okHttp3 logging-interceptor 의존성 추가

kutta97 committed:
retrofitClient 에 okHttpClient 추가함

kutta97 committed:
htmlToString() 함수를 ViewHolder 클래스 안으로 옮기고 private 하게 만들어줌

kutta97 committed:
감독, 출연진 String 에서 | 를 빼주고 , 를 붙여주는 addCommas() 함수 작성

nukeolaf committed:
SearchView 의 xml 속성 namespace 를 android 에서 app 으로 변경

Co-Authored-By: Daeyeon Kim <36095102+improve777@users.noreply.github.com>

kutta97 committed:
MainActivity 에서 searchView 속성관련 코드 제거

kutta97 committed:
Adapter 생성자에서 list 를 제거
list 를 Adapter 의 lateinit var 전역변수로 변경

kutta97 committed:
영화 목록을 ArrayList -> List 로 수정

kutta97 committed:
ViewHolder 의 view 전역변수 제거

kutta97 committed:
Adapter 의 movies 리스트를 lateinit 으로 바꾸니까 에러떠서 수정함
movies 를 List -> ArrayList 로 수정

kutta97 committed:
ViewHolder bind() 에서 사용하지 않는 position 매개변수 파라미터에서 제거

kutta97 committed:
onCreateViewHolder return 할때 괄호가 과하게 들어가서 제거함

kutta97 committed:
영화 리스트를 ArrayList 로 바꾸고, addAll() 로 관리하도록 수정함

kutta97 committed:
인자로 ArrayList 가 아닌 List 를 받게함

kutta97 committed:
response.body() null 체크 추가

kutta97 committed:
Glide dependency 추가 시 annotationProcessor -> kapt 로 수정

kutta97 committed:
textView 에서 텍스트 속성을 tools:text 로 수정

kutta97 committed:
recyclerView height 을 wrap_content -> 0dp 로 바꾸고, weight 을 1로 줌

kutta97 committed:
Serializable 을 Parcelable 로 변경

Android Extensions 플러그인 실험실 기능 사용

kutta97 committed:
영화 재 검색할때마다 movies list clear() 로 지워주도록 함

ddaeyeonkim commented on 2004/nukeolaf/app/src/main/java/com/olaf/nukeolaf/MovieAdapter.kt:
```diff
@@ -0,0 +1,83 @@
+    private val itemListener: MovieItemListener
+) : RecyclerView.Adapter<MovieAdapter.ViewHolder>() {
+
+    private var movies: ArrayList<MovieItem> = ArrayList()
```
`val` 로 변경해주세요

nukeolaf replied:
수정했습니다! a84102b

ddaeyeonkim commented on 2004/nukeolaf/app/build.gradle:
```diff
@@ -21,15 +23,25 @@ android {
             proguardFiles getDefaultProguardFile('proguard-android-optimize.txt'), 'proguard-rules.pro'
         }
     }
+    androidExtensions {
+        experimental = true
+    }
```
`Parcelize` 어노테이션 사용할때 코틀린 1.3.40 버전 이전에만 experimental  이었던 것으로 알고 있습니다.
제거해주셔도 될것 같아요.

nukeolaf replied:
writeToParcel()/createFromParcel() 를 자동으로 구현해주기 위해 experimental 을 사용했어요
experimental 을 제거하면 Parcelize 어노테이션 사용 시 구현해주라고 빨간줄이 떠요ㅠㅠ

ddaeyeonkim replied:
ide 문제인 걸까요..ㅜ 알겠습니다.

TaeseongYun commented on 2004/nukeolaf/app/src/main/java/com/olaf/nukeolaf/MovieAdapter.kt:
```diff
@@ -0,0 +1,83 @@
+                }
+                Glide.with(this)
+                    .load(item.image)
+                    .into(this.movie_image)
```
Glide 처리할 때 error 처리도 하시면 좋을거 같아요! 가끔 영화사진이 없는게 있어요

nukeolaf replied:
수정했습니다! ded7d12

kutta97 committed:
BuildConfig 에 API 키 추가

kutta97 committed:
String resource 에서 api 키 제거

kutta97 committed:
Retrofit Interface 에서 Header 제거
(OkHttp Interceptor 로 header 추가하도록 변경할 것임)

kutta97 committed:
searchMovie 함수에서 header 파라미터 제거

kutta97 committed:
Retrofit client 에 HeaderInterceptor 추가

kutta97 committed:
에러때문에 JVM 타겟 1.8 로 변경
Cannot inline bytecode built with JVM target 1.8 into bytecode that is being built with JVM target 1.6. Please specify proper '-jvm-target' option

kutta97 committed:
movieAdapter 의 movies 리스트를 val 로 수정

kutta97 committed:
Glide 에 error() 추가

ddaeyeonkim reviewed (approved):
LGTM 👍👍👍

sjjeong commented on 2004/nukeolaf/app/src/main/java/com/olaf/nukeolaf/MovieAdapter.kt:
```diff
@@ -0,0 +1,85 @@
+    private val itemListener: MovieItemListener
+) : RecyclerView.Adapter<MovieAdapter.ViewHolder>() {
+
+    private val movies: ArrayList<MovieItem> = ArrayList()
```
Kotlin에서는 ArrayList보다 MutableList를 사용하도록 해요
```kotlin
private val movies = mutableListOf<MovieItem>()
```

nukeolaf replied:
수정했습니다! 
9165f29

그런데 ArrayList 나 MutableList 나 Compile Time 에서는 다 같은 List 로 인식되어서 둘의 큰 차이는 없다고 하는데, 왜 MutableLilst 를 사용하도록 권장하는지, ArrayList 는 어떤경우에 사용하는게 좋은지 여쭤보아도 될까요?

sjjeong replied:
많은 collection이나 exception류 들이 typealias되어 있어요
나중에 어떠한 이유 때문에 ArrayList같은 것을 사용하지 못할때 대응을 해주기 위해서 에요

sjjeong commented on 2004/nukeolaf/app/src/main/java/com/olaf/nukeolaf/MovieAdapter.kt:
```diff
@@ -0,0 +1,85 @@
+        holder.bind(movies[position])
+    }
+
+    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
```
ViewHolder에는 inner를 가급적이면 사용하지 않도록 하는게 좋습니다

nukeolaf replied:
수정했습니다!
984199a

ViewHolder 에 inner 사용을 지양하라고 하시는 이유는 메모리 릭 때문인가요?
찾아보니까 ViewHolder 를 static 으로 사용하는 방법도 있던데 어떤차이인지 궁금합니다

sjjeong replied:
inner class가 static class인건 아닌걸로 알고 있어요
ViewHolder를 inner class로 쓰면 나중에 View pool에서 메모리 릭 문제가 발생해요

sjjeong commented on 2004/nukeolaf/app/src/main/java/com/olaf/nukeolaf/MovieAdapter.kt:
```diff
@@ -0,0 +1,85 @@
+                movie_actor.text = item.actor.addCommas("출연진 : ")
+                movie_rating.numStars = 5
+                movie_rating.rating = item.userRating / 2
+                setOnClickListener {
+                    itemListener.onMovieItemClick(item)
+                }
```
click listener는 onCreateViewHolder 시점인 ViewHolder가 생성되는 시점에 해주면 됩니다.
position은 adapterPosition으로 접근 가능합니다.

nukeolaf replied:
어떤 말씀이신지 잘 이해가 안돼서 이렇게 수정했는데
제가 잘 이해한것이 맞는지 확인 부탁드립니다!
832059f

sjjeong replied:
setOnClickListener 등록하는것을 ViewHolder 생성 시점에서 처리 해주시면 돼요
ViewHolder 클래스가 객체로 만들어지는 시점에요

nukeolaf replied:
수정했습니다!
000f2d0

sjjeong commented on 2004/nukeolaf/app/build.gradle:
```diff
@@ -14,22 +16,41 @@ android {
         versionCode 1
         versionName "1.0"
         testInstrumentationRunner "androidx.test.runner.AndroidJUnitRunner"
+        buildConfigField("String", "Naver_Client_Id", Naver_Client_Id)
+        buildConfigField("String", "Naver_Client_Secret", Naver_Client_Secret)
```
상수는 대문자로 선언해주세요
NAVER_CLIENT_ID
NAVER_CLIENT_SECRET

nukeolaf replied:
수정했습니다!
3b709ec

sjjeong commented on 2004/nukeolaf/app/src/main/java/com/olaf/nukeolaf/RetrofitClient.kt:
```diff
@@ -0,0 +1,40 @@
+            .addInterceptor { chain: Interceptor.Chain ->
+                val original = chain.request()
+                chain.proceed(original.newBuilder().apply {
+                    addHeader("X-Naver-Client-Id", Naver_Client_Id)
+                    addHeader("X-Naver-Client-Secret", Naver_Client_Secret)
```
이렇게 사용하면 Naver_Client_Id가 어디에 있는 값인지 알기가 어렵습니다.
BuildConfig.Naver_Client_Id 처럼 사용해주셔야 합니다.

nukeolaf replied:
수정했습니다!
3cbf989

```
import com.olaf.nukeolaf.BuildConfig.Naver_Client_Id
import com.olaf.nukeolaf.BuildConfig.Naver_Client_Secret
```
이렇게 import 되어 있어서 알 수 있다고 생각했는데,
import 로 어디에 있는 값인지 확인하는 것은 이상한가요??

sjjeong replied:
코드 가독성을 높이기 위해서에요.
사용하는쪽에서 어떤 클래스에서 있는 상수인지 바로 알면 가독성이 올라가는데
그렇지 않으면 이 상수가 어디서 오는지 찾아봐야하는 불편함이 있어요

sjjeong commented on 2004/nukeolaf/app/src/main/java/com/olaf/nukeolaf/RetrofitClient.kt:
```diff
@@ -0,0 +1,40 @@
+                }.build())
+            }
+            .addInterceptor(HttpLoggingInterceptor().apply {
+                level = HttpLoggingInterceptor.Level.BODY
```
BuildConfig.DEBUG가 true일 때 BODY를 넣고 false이면 NONE을 넣도록 해주세요

nukeolaf replied:
수정했습니다!
5ed8854

sjjeong replied:
if else가 한 라인에 들어가지 않는다면 {}를 추가해주셔야 해요

nukeolaf replied:
수정했습니다!
5433c4d

sjjeong commented on 2004/nukeolaf/app/src/main/res/layout/activity_main.xml:
```diff
@@ -1,18 +1,25 @@
+    <androidx.recyclerview.widget.RecyclerView
+        android:id="@+id/movie_rv"
+        android:layout_width="match_parent"
+        android:layout_height="0dp"
+        android:layout_weight="1"
```
지금은 weight가 불필요해 보이네요
height를 match로 해도 됩니다.

nukeolaf replied:
수정했습니다!
9007b60

sjjeong commented on 2004/nukeolaf/app/src/main/java/com/olaf/nukeolaf/MainActivity.kt:
```diff
@@ -1,12 +1,82 @@
 
 class MainActivity : AppCompatActivity() {
 
+    private lateinit var retrofitClient: RetrofitInterface
```
lateinit var 말고 val로 선언해서 바로 가져와도 될 것 같습니다.
```kotlin
private val retrofitClient: RetrofitInterface = RetrofitClient.client
```

nukeolaf replied:
수정했습니다!
d043ab2

sjjeong commented on 2004/nukeolaf/app/src/main/java/com/olaf/nukeolaf/MainActivity.kt:
```diff
@@ -1,12 +1,82 @@
+
+        search_view.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
+            override fun onQueryTextSubmit(query: String?): Boolean {
+                searchMovie(query!!)
```
!!는 사용하지 않는게 좋습니다.
null인지 검사해서 null이 아니면 searchMovie를 호출하도록 해주세요

nukeolaf replied:
수정했습니다!
0ca2c40

아무것도 입력 안하고 검색을 눌러도 null 일때 코드가 실행되지 않아서 !! 로 바꿨는데
!! 는 그럼 어떤 상황에 사용하는것이 좋을까요?

sjjeong replied:
어떤 경우에도 사용하지 않는게 좋아요
!!는 NPE를 발생시켜서 앱이 중지되는 문제가 있어요

sjjeong commented on 2004/nukeolaf/app/src/main/java/com/olaf/nukeolaf/MainActivity.kt:
```diff
@@ -1,12 +1,82 @@
+                call: Call<MovieResponse>,
+                response: Response<MovieResponse>
+            ) {
+                if (response.body() != null && response.isSuccessful) {
+                    val movies = response.body()!!.items
```
response.body()가 null이 아닌 것을 체크 했는데 nullable이 유지되고 있습니다.
이 코드는 아래처럼 변경하면 해결 됩니다.
```
val body = response.body()
if (body != null && response.isSuccessful) {
    val movies = body.items
```

nukeolaf replied:
수정했습니다!
803af4b

sjjeong commented on 2004/nukeolaf/app/src/main/java/com/olaf/nukeolaf/MainActivity.kt:
```diff
@@ -1,12 +1,82 @@
+    private fun showMovies(movies: List<MovieItem>) {
+        movieAdapter.run {
+            setMovies(movies)
+            notifyDataSetChanged()
```
notifyDataSetChanged()를 setMovies 함수 내부로 이동하면 아래처럼 사용 할 수 있습니다.
```kotlin
movieAdapter.setMovies(movies)
```

nukeolaf replied:
수정했습니다!
915505d

kutta97 committed:
영화 목록 ArrayList -> mutableListOf 로 수정

kutta97 committed:
ViewHolder 를 외부 클래스로 분리

kutta97 committed:
ViewHolder 생성자에 아이템 리스너 추가

kutta97 committed:
네이버 API 키 관련 상수 대문자로 수정

kutta97 committed:
네이버 API 키 상수 앞에 BuildConfig. 붙여서 사용하도록 수정

kutta97 committed:
BuildConfig.DEBUG 가 true 일때만 로깅 인터셉터 Body 를 넣고, 아닐때는 None 으로 수정

kutta97 committed:
리사이클러뷰 레이아웃 height 을 match_parent 로 수정하고 weight 삭제

kutta97 committed:
MainActivity 에서 retrofitClient lateinit var -> val 로 수정

kutta97 committed:
SearchView 로 영화 검색시 query null 검사하는 코드 추가

kutta97 committed:
response.body null 검사 코드 수정

kutta97 committed:
notifyDataSeyChanged() 를 setMovies 함수 안으로 이동

kutta97 committed:
리사이클러뷰 아이템뷰 클릭 리스너를 ViewHolder 생성 시점으로 변경

kutta97 committed:
if else 구문에 {} 추가

sjjeong reviewed (approved):
LGTM ✊ ✌️ 👋
